snap/validate.go

```diff
@@ -53,13 +54,53 @@ func ValidateName(name string) error {
 // NB keep this in sync with snapcraft and the review tools :-)
 var isValidVersion = regexp.MustCompile("^[a-zA-Z0-9](?:[a-zA-Z0-9:.+~_-]{0,30}[a-zA-Z0-9+~])?$").MatchString
 
+var isInvalidFirstVersionChar = regexp.MustCompile("^[^a-zA-Z0-9]").MatchString
+var isInvalidLastVersionChar = regexp.MustCompile("[^a-zA-Z0-9+~]$").MatchString
+var invalidMiddleVersionChars = regexp.MustCompile("[^a-zA-Z0-9:.+~_-]+").FindAllString
+
 // ValidateVersion checks if a string is a valid snap version.
 func ValidateVersion(version string) error {
-	if len(version) == 0 {
-		return fmt.Errorf("snap version cannot be empty")
-	}
 	if !isValidVersion(version) {
-		return fmt.Errorf("invalid snap version: %q", version)
+		// maybe it was too short?
+		if len(version) == 0 {
+			return fmt.Errorf("invalid snap version: cannot be empty")
+		}
+		if !strutil.IsPrintableASCII(version) {
+			return fmt.Errorf("invalid snap version %s: must be printable, non-whitespace ASCII",
+				strconv.QuoteToASCII(version))
+		}
+		// now we know it's a non-empty ASCII string, we can get serious
+		var reasons []string
+		// ... too long?
+		if len(version) > 32 {
+			reasons = append(reasons, fmt.Sprintf("cannot be longer than 32 characters (got: %d)", len(version)))
+		}
+		// started with a symbol?
+		if isInvalidFirstVersionChar(version) {
+			// note that we can only say version[0] because we know it's ASCII :-)
+			reasons = append(reasons, fmt.Sprintf("must start with an ASCII alphanumeric (and not %q)", version[0]))
+		}
+		if len(version) > 1 {
```

Comment:
You check for empty version up front, you don't need this check here IMO

Reply:
`>0` != `>1`.

Reply:
(when `len(s) == 1`, you can do `s[1:]`, you can do `s[:len(s)-1]`, but you can't do `s[1:len(s)-1]`.)